lib/kafka/offset_manager.rb

```diff
@@ -62,9 +62,15 @@ def commit_offsets_if_necessary
       end
     end
 
-    def clear_offsets
-      @uncommitted_offsets = 0
-      @processed_offsets.clear
+    def clear_offsets_excluding(excluded)
+      # Clear all offsets that aren't in `excluded`.
+      @processed_offsets.each do |topic, partitions|
+        partitions.keep_if do |partition, _|
+          excluded.fetch(topic, []).include?(partition)
+        end
+      end
+
+      # Clear the cached commits from the brokers.
       @committed_offsets = nil
```

Comment:
`@processed_offsets` refers to the consumers in-memory checkpoints, updated after _every_ message. `@committed_offsets` is a cache of what the broker API returned, i.e. the _committed_ offsets.